packages/contracts-bedrock/src/L2/SuperchainERC20Factory.sol

```diff
@@ -0,0 +1,186 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.25;
+
+import { ISemver } from "src/universal/ISemver.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+import { IL2ToL2CrossDomainMessenger } from "src/L2/IL2ToL2CrossDomainMessenger.sol";
+import { ICrossDomainMessenger } from "@openzeppelin/contracts/vendor/optimism/ICrossDomainMessenger.sol";
+import { BeaconProxy } from "@openzeppelin/contracts/proxy/beacon/BeaconProxy.sol";
+import { CREATE3 } from "@rari-capital/solmate/src/utils/CREATE3.sol";
+
+/// @notice Thrown when attempting to deploy a SuperchainERC20 from L1 and the caller is not the L2CrossDomainMessenger.
+error CallerNotL2CrossDomainMessenger();
+
+/// @notice Thrown when attempting to deploy a SuperchainERC20 from L1 and the xDomainMessageSender is not the native
+/// initializer.
+error InvalidNativeInitializer();
+
+/// @notice Thrown when attempting to deploy a SuperchainERC20 from L2 and the caller is not the
+/// L2ToL2CrossDomainMessenger.
+error CallerNotL2ToL2CrossDomainMessenger();
+
+/// @notice Thrown when attempting to deploy a SuperchainERC20 from L2 and the crossDomainMessageSender is not the
+/// SuperchainERC20Factory
+error InvalidCrossDomainSender();
+
+/// @notice Thrown when attempting to deploy a SuperchainERC20 and the deployment already exists.
+error AlreadyDeployed(address _remoteToken);
+
+/// @notice Thrown when attempting to deploy a SuperchainERC20 with metadata and the deployment does not exist.
+error NotDeployed();
+
+/// @custom:proxied
+/// @title SuperchainERC20Factory
+/// @notice SuperchainERC20Factory is a factory contract that deploys SuperchainERC20 Beacon Proxies using CREATE3.
+contract SuperchainERC20Factory is ISemver {
+    /// TODO: Replace with real native initializer address
+    /// @notice Address of the native initializer.
+    address internal constant NATIVE_INITIALIZER = 0x0000000000000000000000000000000000000000;
+
+    /// @notice Address of the L2CrossDomainMessenger Predeploy.
+    address internal constant L2_CROSS_DOMAIN_MESSENGER = Predeploys.L2_CROSS_DOMAIN_MESSENGER;
```

Comment:
we probably can skip the declaration of all these variables and use `Predeploys.L2_CROSS_DOMAIN_MESSENGER` directly, or even import the specific variables from the `constants` file